File: apps/go/internal/chapter10/featureimportance.go
package chapter10

import (
	"fmt"

	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter02"
	"github.com/k2works/getting-started-machine-learning/apps/go/internal/chapter03"
)

// Importance は特徴量と、その重要度。列の順に並べて使う。
type Importance struct {
	Feature string
	Value   float64
}

// TreeImportances は決定木 1 本の重要度を、特徴量の列の順に並べて返す。合計は 1 になる。
func TreeImportances(tree chapter03.Tree, x []chapter02.Features, t []string) ([]Importance, error) {
	totals := make(map[string]float64, len(x[0].Columns))

	if err := addDecreases(tree, x, t, totals); err != nil {
		return nil, err
	}

	return normalize(x[0].Columns, totals), nil
}

// TreeImportancesOf は深さを制限しない決定木を学習してから、その重要度を求める。
func TreeImportancesOf(x []chapter02.Features, t []string) ([]Importance, error) {
	model := chapter03.Unlimited()
	if err := model.Fit(x, t); err != nil {
		return nil, err
	}

	tree, ok := model.Tree()
	if !ok {
		return nil, fmt.Errorf("決定木がまだ作られていません")
	}

	return TreeImportances(tree, x, t)
}

// addDecreases は分割で減った不純度（件数で重み付け）を、特徴量ごとに足し込む。
func addDecreases(tree chapter03.Tree, x []chapter02.Features, t []string, totals map[string]float64) error {
	node, ok := tree.(chapter03.Node)
	if !ok {
		// 葉には分割が無いので、減った不純度も無い
		return nil
	}

	var leftX, rightX []chapter02.Features

	var leftT, rightT []string

	for i, features := range x {
		value, err := features.Value(node.Split.Feature)
		if err != nil {
			return err
		}

		if value <= node.Split.Threshold {
			leftX, leftT = append(leftX, features), append(leftT, t[i])
		} else {
			rightX, rightT = append(rightX, features), append(rightT, t[i])
		}
	}

	totals[node.Split.Feature] += float64(len(t)) * (chapter03.Gini(t) - node.Split.Impurity)

	if err := addDecreases(node.Left, leftX, leftT, totals); err != nil {
		return err
	}

	return addDecreases(node.Right, rightX, rightT, totals)
}

// normalize は合計が 1 になるようにそろえ、列の順に並べて返す。合計が 0 ならそのまま返す。
func normalize(columns []string, totals map[string]float64) []Importance {
	sum := 0.0
	for _, value := range totals {
		sum += value
	}

	importances := make([]Importance, len(columns))

	for i, column := range columns {
		value := totals[column]
		if sum != 0 {
			value /= sum
		}

		importances[i] = Importance{Feature: column, Value: value}
	}

	return importances
}

// ForestImportances は木ごとに正規化した重要度の平均を返す。木が使わなかった特徴量は、その木では 0 とする。
func ForestImportances(forest *RandomForest, x []chapter02.Features, t []string) ([]Importance, error) {
	if len(forest.Trees()) == 0 {
		return nil, fmt.Errorf("Fit で学習してから ForestImportances を呼んでください")
	}

	totals := make(map[string]float64, len(x[0].Columns))

	for _, fitted := range forest.Trees() {
		sampleX := make([]chapter02.Features, len(fitted.Rows))
		sampleT := make([]string, len(fitted.Rows))

		for i, row := range fitted.Rows {
			sampleX[i] = x[row]
			sampleT[i] = t[row]
		}

		selected, err := SelectColumns(sampleX, fitted.Columns)
		if err != nil {
			return nil, err
		}

		tree, ok := fitted.Model.Tree()
		if !ok {
			return nil, fmt.Errorf("決定木がまだ作られていません")
		}

		importances, err := TreeImportances(tree, selected, sampleT)
		if err != nil {
			return nil, err
		}

		for _, importance := range importances {
			totals[importance.Feature] += importance.Value / float64(len(forest.Trees()))
		}
	}

	return normalize(x[0].Columns, totals), nil
}

File: apps/csharp/src/MachineLearning/Chapter10/FeatureImportance.cs
namespace MachineLearning.Chapter10;

using MachineLearning.Chapter02;
using MachineLearning.Chapter03;

/// <summary>決定木とランダムフォレストの特徴量の重要度。</summary>
public static class FeatureImportance
{
    /// <summary>決定木 1 本の特徴量の重要度。</summary>
    public static IReadOnlyDictionary<string, double> TreeImportances(
        Tree tree, IReadOnlyList<Features> x, IReadOnlyList<string> t)
    {
        ArgumentNullException.ThrowIfNull(x);
        var decreases = new Dictionary<string, double>(StringComparer.Ordinal);
        Collect(tree, x, t, decreases);
        return Normalize(x[0].Columns.ToDictionary(
            column => column, decreases.GetValueOrDefault, StringComparer.Ordinal));
    }

    /// <summary>
    /// ランダムフォレストの特徴量の重要度。
    /// 木ごとの重要度（学習に使ったブートストラップ標本で計算）を平均し、割合にする。
    /// </summary>
    public static IReadOnlyDictionary<string, double> ForestImportances(
        IReadOnlyList<FittedTree> forest, IReadOnlyList<Features> x, IReadOnlyList<string> t)
    {
        ArgumentNullException.ThrowIfNull(forest);
        ArgumentNullException.ThrowIfNull(x);
        var perTree = forest.Select(fitted =>
        {
            var (sampleX, sampleT) = RandomForest.SampleOf(fitted.Rows, fitted.Columns, x, t);
            return TreeImportances(fitted.Tree, sampleX, sampleT);
        }).ToList();

        return Normalize(x[0].Columns.ToDictionary(
            column => column,
            column => perTree.Average(importances => importances.GetValueOrDefault(column)),
            StringComparer.Ordinal));
    }

    /// <summary>
    /// 学習に使ったデータをもう一度木に流して、
    /// 節ごとに「分割に使った特徴量と、件数で重み付けした不純度の減少量」を集める。
    /// </summary>
    private static void Collect(
        Tree tree, IReadOnlyList<Features> x, IReadOnlyList<string> t, Dictionary<string, double> decreases)
    {
        if (tree is not Node node)
        {
            return;
        }

        var split = node.Split;
        var left = Enumerable.Range(0, x.Count).Where(i => x[i].Value(split.Feature) <= split.Threshold).ToList();
        var right = Enumerable.Range(0, x.Count).Where(i => x[i].Value(split.Feature) > split.Threshold).ToList();
        decreases[split.Feature] =
            decreases.GetValueOrDefault(split.Feature) + (t.Count * (DecisionTrees.Gini(t) - split.Impurity));
        Collect(node.Left, Pick(x, left), Pick(t, left), decreases);
        Collect(node.Right, Pick(x, right), Pick(t, right), decreases);
    }

    /// <summary>合計が 1 になるように割合にする。合計が 0 ならそのまま返す。</summary>
    private static Dictionary<string, double> Normalize(Dictionary<string, double> totals)
    {
        var total = totals.Values.Sum();
        return total == 0.0
            ? totals
            : totals.ToDictionary(pair => pair.Key, pair => pair.Value / total, StringComparer.Ordinal);
    }

    private static IReadOnlyList<T> Pick<T>(IReadOnlyList<T> values, IReadOnlyList<int> positions) =>
        [.. positions.Select(i => values[i])];
}
